### ISSUE TITLE
[Android 8 and lower] ANR crash detection broken + breaks entire plugin when enabled

### ISSUE THREAD

**Issue by vlussenburg** [backtrace-labs/backtrace-unity#27]
```07-02 16:18:22.840 19267 19279 E Unity   : FileNotFoundException: Could not find file "/Path to not existing file"
07-02 16:18:22.840 19267 19279 E Unity   :   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x001b7] in <2fa7a6a452ca43df998f07fd1486c0df>:0 
07-02 16:18:22.840 19267 19279 E Unity   :   at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share) [0x00000] in <2fa7a6a452ca43df998f07fd1486c0df>:0 
07-02 16:18:22.840 19267 19279 E Unity   :   at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare)
07-02 16:18:22.840 19267 19279 E Unity   :   at System.IO.File.OpenRead (System.String path) [0x00000] in <2fa7a6a452ca43df998f07fd1486c0df>:0 
07-02 16:18:22.840 19267 19279 E Unity   :   at System.IO.File.ReadAllBytes (System.String path) [0x00000] in <2fa7a6a452ca43df998f07fd1486c0df>:0 
07-02 16:18:22.840 19267 19279 E Unity   :   at PlayerShip.Fire () [0x000c1] in <96b16c5b85c14628918aa833cc8a7a29>:0 
07-02 16:18:22.840 19267 19279 E Unity   :   at PlayerShip.Update () [0x0005f] in <96b16c5b85c14628918aa833cc8a7a29>:0 
07-02 16:18:22.840 19267 19279 E Unity   :  
07-02 16:18:22.840 19267 19279 E Unity   : (F
07-02 16:18:22.920 19267 19279 E Unity   : NullReferenceException: Object reference not set to an instance of an object
07-02 16:18:22.920 19267 19279 E Unity   :   at Backtrace.Unity.BacktraceClient.HandleUnityMessage (System.String message, System.String stackTrace, UnityEngine.LogType type) [0x00009] in <84e72275204c4bf38ff2688ac78f066e>:0 
07-02 16:18:22.920 19267 19279 E Unity   :   at UnityEngine.Application.CallLogCallback (System.String logString, System.String stackTrace, UnityEngine.LogType type, System.Boolean invokedOnMainThread) [0x00014] in <d71499eab75645c19df09ba17af9a506>:0 
```

See `NullReferenceException`.

I believe the root cause is:

```
07-02 16:45:51.900 31827 31855 E Unity   : AndroidJavaException: java.lang.NoSuchMethodError: no non-static method "Lbacktrace/io/backtrace_unity_android_plugin/BacktraceANRWatchdog;.<init>()V"
07-02 16:45:51.900 31827 31855 E Unity   : java.lang.NoSuchMethodError: no non-static method "Lbacktrace/io/backtrace_unity_android_plugin/BacktraceANRWatchdog;.<init>()V"
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer.nativeRender(Native Method)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer.c(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer$e$2.queueIdle(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   : 	at android.os.MessageQueue.next(MessageQueue.java:392)
07-02 16:45:51.900 31827 31855 E Unity   : 	at android.os.Looper.loop(Looper.java:143)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer$e.run(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine.AndroidJNISafe.CheckException () [0x00091] in <d71499eab75645c19df09ba17af9a506>:0 
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine.AndroidJNISafe.GetMethodID (System.IntPtr obj, System.String name, System.String sig) [0x00011] in <d71499eab75645c19df09ba17af9a506>:0 
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine._AndroidJNIHelper.GetConstructorID (System.IntPtr jclass, System.String signature) [0x0001e] in <d71499eab75645c19df09ba17af9a506>:0 
```

**Comment by vlussenburg** [comment]
Another similar or same problem:

```07-02 16:18:23.240 19267 19279 E Unity   : NullReferenceException: Object reference not set to an instance of an object
07-02 16:18:23.240 19267 19279 E Unity   :   at Backtrace.Unity.BacktraceClient.Send (Backtrace.Unity.Model.BacktraceReport report, System.Action`1[T] sendCallback) [0x0000a] in <84e72275204c4bf38ff2688ac78f066e>:0 
07-02 16:18:23.240 19267 19279 E Unity   :   at PlayerShip.Fire () [0x000ac] in <96b16c5b85c14628918aa833cc8a7a29>:0 
07-02 16:18:23.240 19267 19279 E Unity   :   at PlayerShip.Update () [0x0005f] in <96b16c5b85c14628918aa833cc8a7a29>:0 
07-02 16:18:23.240 19267 19279 E Unity   :  
07-02 16:18:23.240 19267 19279 E Unity   : (Filename: <84e72275204c4bf38ff2688ac78f066e> Line: 0)
07-02 16:18:23.240 19267 19279 E Unity   : 
```

**Comment by vlussenburg** [comment]
Based on these two traces I suspect `_backtraceLogManager.Enqueue(report);`

**Comment by vlussenburg** [comment]
I believe the root cause is:

```
07-02 16:45:51.900 31827 31855 E Unity   : AndroidJavaException: java.lang.NoSuchMethodError: no non-static method "Lbacktrace/io/backtrace_unity_android_plugin/BacktraceANRWatchdog;.<init>()V"
07-02 16:45:51.900 31827 31855 E Unity   : java.lang.NoSuchMethodError: no non-static method "Lbacktrace/io/backtrace_unity_android_plugin/BacktraceANRWatchdog;.<init>()V"
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer.nativeRender(Native Method)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer.c(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer$e$2.queueIdle(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   : 	at android.os.MessageQueue.next(MessageQueue.java:392)
07-02 16:45:51.900 31827 31855 E Unity   : 	at android.os.Looper.loop(Looper.java:143)
07-02 16:45:51.900 31827 31855 E Unity   : 	at com.unity3d.player.UnityPlayer$e.run(Unknown Source)
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine.AndroidJNISafe.CheckException () [0x00091] in <d71499eab75645c19df09ba17af9a506>:0 
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine.AndroidJNISafe.GetMethodID (System.IntPtr obj, System.String name, System.String sig) [0x00011] in <d71499eab75645c19df09ba17af9a506>:0 
07-02 16:45:51.900 31827 31855 E Unity   :   at UnityEngine._AndroidJNIHelper.GetConstructorID (System.IntPtr jclass, System.String signature) [0x0001e] in <d71499eab75645c19df09ba17af9a506>:0 
```

**Comment by vlussenburg** [comment]
```
        public void Refresh()
        {
            Debug.Log("here1");
            if (Configuration == null || !Configuration.IsValid())
            {
                return;
            }

            Enabled = true;
            Debug.Log("here2");

            // set maximum game object depth
            _gameObjectDepth = Configuration.GameObjectDepth == 0
                ? 16 // default maximum game object size
                : Configuration.GameObjectDepth;
            Debug.Log("here3");
            CaptureUnityMessages();
            _reportLimitWatcher = new ReportLimitWatcher(Convert.ToUInt32(Configuration.ReportPerMin));
            Debug.Log("here4");

            _nativeClient = NativeClientFactory.GetNativeClient(Configuration, name);
            Debug.Log("here5");
```

Doesn't get to here5 in `BactraceClient.cs`

**Comment by vlussenburg** [comment]
Confirmed: Backtrace client works again when I disable ANR in the configuration options.

**Comment by vlussenburg** [comment]
Device info:

```json
{
  "system.memory.swap.free": "1416516",
  "system.memory.active": "790904",
  "vm.sigpnd": "0000000000000000",
  "app.storage_used": "8643544",
  "system.memory.vmalloc.chunk": "247561408",
  "system.memory.cached": "994952",
  "system.memory.swap.total": "2097148",
  "device.sdk": "23",
  "vm.rss.peak": "362520",
  "system.memory.free": "48188",
  "vm.rss.size": "344340",
  "vm.vma.peak": "1548452",
  "vm.shared.size": "123584",
  "culture": "English",
  "device.is_power_saving_mode": "false",
  "system.memory.slab": "134568",
  "vm.locked.size": "0",
  "system.memory.swap.cached": "142508",
  "device.cpu.temperature": "59.0",
  "vm.pte.size": "1232",
  "sched.cs.voluntary": "336",
  "vm.shdpnd": "0000000000000000",
  "descriptor.count": "128",
  "device.manufacturer": "samsung",
  "system.memory.writeback": "0",
  "vm.swap.size": "6168",
  "vm.vma.size": "1430716",
  "system.memory.vmalloc.used": "269204",
  "system.memory.buffers": "34156",
  "vm.exe": "20",
  "system.memory.vmalloc.total": "251658176",
  "sched.cs.involuntary": "65",
  "state": "S (sleeping)",
  "vm.threads": "60",
  "vm.data": "311828",
  "system.memory.dirty": "1124",
  "system.memory.total": "2744140",
  "system.memory.inactive": "780784",
  "vm.stack.size": "8192",
  "error.message": "InsufficientMemoryException: Insuff mem.",
  "guid": "b13c566f6ba221c9aab0c5de42c2e186",
  "scripting.backend": ".NET Standard 2.0",
  "application": "AsteraX",
  "application.version": "3.0.0",
  "application.url": "",
  "application.company.name": "Backtrace.io",
  "application.data_path": "/data/app/io.backtrace.asterax-1/base.apk",
  "application.id": "io.backtrace.asterax",
  "application.installer.name": "",
  "application.internet_reachability": "ReachableViaLocalAreaNetwork",
  "application.editor": "False",
  "application.focused": "True",
  "application.mobile": "True",
  "application.playing": "True",
  "application.background": "False",
  "application.sandboxType": "Sandboxed",
  "application.system.language": "English",
  "application.unity.version": "2018.4.23f1",
  "application.temporary_cache": "/storage/emulated/0/Android/data/io.backtrace.asterax/cache",
  "application.debug": "False",
  "battery.level": "55",
  "battery.status": "Charging",
  "device.model": "samsung SM-G920W8",
  "device.name": "Galaxy S6",
  "device.type": "Handheld",
  "graphic.id": "0",
  "graphic.name": "Mali-T760",
  "graphic.type": "OpenGLES3",
  "graphic.vendor": "ARM",
  "graphic.vendor.id": "0",
  "graphic.driver.version": "OpenGL ES 3.1 v1.r7p0-03rel0.b596bd02e7d0169c10574b57180c8b57",
  "graphic.memory": "1024",
  "graphic.multithreaded": "True",
  "graphic.shader": "50",
  "graphic.topUv": "False",
  "uname.sysname": "Android",
  "uname.version": "3.10.61.9889024",
  "uname.fullname": "Android OS 6.0.1 / API-23 (MMB29K/G920W8VLU5CPK4)",
  "uname.family": "Other",
  "cpu.count": "8",
  "cpu.frequency": "2100",
  "cpu.brand": "ARMv7 VFPv3 NEON",
  "audio.supported": "True",
  "cpu.boottime": "7116563",
  "hostname": "localhost",
  "gc.heap.used": "1277952",
  "process.age": "11",
  "scene.count.build": "1",
  "scene.count": "1",
  "scene.active": "_Scene_0",
  "scene.buildIndex": "0",
  "scene.handle": "-70",
  "scene.isDirty": "False",
  "scene.isLoaded": "True",
  "scene.name": "_Scene_0",
  "scene.path": "Assets/_Scenes/_Scene_0.unity"
}
```

**Comment by vlussenburg** [comment]
Update: fails on emulator too. Can't work AFAIK.

This:
```java
_anrWatcher = new AndroidJavaObject(_anrPath);
_anrWatcher.CallStatic("watch", gameObjectName, callbackName);
```

Needs to be be:

```java
new AndroidJavaClass(_anrPath).CallStatic("watch", gameObjectName, callbackName);
```

(I removed the `anrWatcher` which wasn't used anyways), but you could add that using [AndroidJavaClass](https://docs.unity3d.com/ScriptReference/AndroidJavaClass.html)`#GetStatic` on the `_instance`.

**Comment by vlussenburg** [comment]
Tested
[x] Android 10 (emulator)
[x] Android 6 (device)
[ ] Android 7 (device)

**Comment by vlussenburg** [comment]
Fails on Android 6-8, runs fine on 10, didn't test 9 (don't have a test image).

**Comment by konraddysput** [comment]
@vlussenburg I think bug is somewhere else - I will download today emulators and try to recognize what is wrong.

**Comment by konraddysput** [comment]
static method `watch` is responsible for creating a new BacktraceANRWatchdog object - that is responsible for detecting possible ANRs. `watch` method *shouldn't* require you to create a new instance of an object. 


@vlussenburg can you try to build Untiy game from Unity instead? I saw you tried to install unity game from Android Studio.

**Comment by konraddysput** [comment]
lease see Backtrace-Unity library changes in the diff available here:
https://github.com/backtrace-labs/backtrace-unity/pull/34

**Comment by konraddysput** [comment]
Fix will be available in version 3.0.1 (already on master)

**Comment by vlussenburg** [comment]
Tested master @ 9f298881bb070a7da8ac748a0fe701b45becc4c8

- [x]  Windows Standalone
- [x]  MacOS Editor
- [ ]  MacOs Standalone
- [x]  iOS
- [x]  Android 7 (real device)
- [x] Android 8-10 (emulator)
- [x]  WebGL
